seahorn/lib/linked_list_helper.c

```diff
@@ -64,89 +27,259 @@ static inline void init_short_aws_linked_list(struct aws_linked_list *list,
   }
 }
 
-void sea_nd_init_aws_linked_list_from_front(struct aws_linked_list *list,
-                                            size_t *length) {
+// if nd_voidp() is assigned to next, prev without assigning them sane
+// values, then compiler treats them as undefined.
+void initNode(struct aws_linked_list_node *node) {
+  assume(node);
+  struct aws_linked_list_node *dummy =
+      malloc(sizeof(struct aws_linked_list_node));
+  node->next = dummy;
+  node->prev = dummy;
+}
+
+void sea_nd_init_aws_linked_list_from_head(struct aws_linked_list *list,
+                                           size_t *length) {
+  list->head.prev = NULL;
+  list->tail.next = NULL;
+
   size_t nd_len = nd_size_t();
   *length = nd_len;
   if (nd_len <= 2) {
     init_short_aws_linked_list(list, nd_len);
   } else {
     // HEAD <--> front <--> front_next --> nd ... nd <-- TAIL
-    struct aws_linked_list_node *front = malloc(sizeof(struct aws_linked_list_node));
-    struct aws_linked_list_node *front_next = malloc(sizeof(struct aws_linked_list_node));
+    struct aws_linked_list_node *front =
+        malloc(sizeof(struct aws_linked_list_node));
+    initNode(front);
+    struct aws_linked_list_node *front_next =
+        malloc(sizeof(struct aws_linked_list_node));
+    initNode(front_next);
     aws_linked_list_attach_after(&list->head, front, true);
```

Comment:
before `aws_linked_list_attach_after`s:
node_a<--front-->node_a
node_b<--front_next-->node_b
after `aws_linked_list_attach_after`s:
head<-->front<-->front_next-->nd_ptr ... nd_ptr<---tail
so if wouldn't `initNode`calls be redundant here?

Reply:
thanks Xiang. 

You are correct. I changed init_node to just nullify the next, prev members. I think this is good discipline in case we make a mistake in connecting the links in some way.

Reply:
this is for preventing the next, prev members to be interpreted as `undefined` right?

Reply:
yes.